Wip/grant permissions

matthiasclasen opened the pull request:
I gave this a quick try. Let me know if this is going in the right direction.

matthiasclasen committed:
Fix document creation

There was a missing NULL termination for varargs that got
in the way of creating XdpDocument instances.

matthiasclasen committed:
Implement GrantPermissions

matthiasclasen committed:
Add the api to the header

This will be used to set up initial permissions in
the OpenDocument call later.

matthiasclasen committed:
Fix up parameter handling

To get a strv, you need to use ^as. Avoid leaking memory
at the same time, by generous use of &.

alexlarsson commented:
I think we want to do the actual saving of changes to the database asynchronously to avoid blocking the mainloop for handling other users.

matthiasclasen committed:
Save permissions asynchronously

We don't want to block the mainloop until the database is done
with its business. Also, there is no need at all to save document
here, the list of permissions is reconstructed anyway when loading
the document from the database.

matthiasclasen committed:
Fix creation of XdpPermissions objects

Another instance of missing NULL-terminated varargs.

matthiasclasen committed:
Return cookie from GrantPermission

Eventually, we may want to allow calls to remove permissions
they added. To make that possible, always add a separate permissions
object for each call, and return its unique ID. To make this
work, we have to actually wait until the object is written
to disk before returning.

matthiasclasen committed:
Forgotten file
